pyrate/core/dem_error.py

```diff
@@ -0,0 +1,307 @@
+    # 1. read orbital state vectors and start/stop times from mli.par
+    # 2. for each pixel get the corresponding radar row (from matrix az)
+    # 3. get the corresponding radar time for that row (using linear interpolation)
+    # 4. calculate the satellite XYZ position for that time by interpolating the time and velocity state vectors
+
+    # calc azimuth angle using Vincenty's equations
+    azimuth_angle = np.empty(lat.shape) # pre-allocate 2D numpy array
+    azimuth_angle[:] = np.nan
+    for ix, iy in np.ndindex(lat.shape):
+        if not isnan(sat_lat[ix, iy]):
+            az12 = vincinv(lat[ix, iy], lon[ix, iy], sat_lat[ix, iy], sat_lon[ix, iy], a, b)
+            azimuth_angle[ix, iy] = az12
+    np.reshape(azimuth_angle, lat.shape)
+
+    # old code: spherical azimuth angle calculation
+    #azimuth_angle2 = np.arccos(np.divide(np.multiply(np.sin(sat_lat), np.cos(lat)) - \
+    #                                np.multiply(np.multiply(np.cos(sat_lat), np.sin(lat)), np.cos(sat_lon - lon)), \
+    #                                np.sin(epsilon)))
+    #azimuth_angle_diff = azimuth_angle2 - azimuth_angle
+    #print(np.nanmin(azimuth_angle_diff), np.nanmax(azimuth_angle_diff))
+    # the difference between Vincety's azimuth calculation and the spherical approximation is ~0.001 radians
+
+    # for validation with GAMMA output only
+    azimuth_angle_gamma = -(azimuth_angle - np.pi / 2) # local heading towards satellite
+    # maximum differences to the GAMMA-derived local azimuth angles for Sentinel-1 test data are within +/-0.5 deg
+    # this could be improved by using orbital state vectors to calculate that satellite positions (see above comment)
+
+    # save angles to npy arrays
+    look_angle_file = os.path.join(params[cf.OUT_DIR],'look_angle.npy')
+    np.save(file=look_angle_file, arr=look_angle)
+
+    # save angles to geotiff
+    gt, md, wkt = shared.get_geotiff_header_info(ifg_path)
+    md[ifc.EPOCH_DATE] = None # needs to have a value in write_output_geotiff
+    look_angle_file = os.path.join(params[cf.OUT_DIR], 'look_angle.tif')
+    shared.write_output_geotiff(md, gt, wkt, look_angle, look_angle_file, np.nan)
+    incidence_angle_file = os.path.join(params[cf.OUT_DIR], 'incidence_angle.tif')
+    shared.write_output_geotiff(md, gt, wkt, incidence_angle, incidence_angle_file, np.nan)
+    azimuth_angle_file = os.path.join(params[cf.OUT_DIR], 'azimuth_angle.tif')
+    shared.write_output_geotiff(md, gt, wkt, azimuth_angle, azimuth_angle_file, np.nan)
+    incidence_angle_gamma_file = os.path.join(params[cf.OUT_DIR], 'incidence_angle_gamma.tif')
+    shared.write_output_geotiff(md, gt, wkt, incidence_angle_gamma, incidence_angle_gamma_file, np.nan)
+    azimuth_angle_gamma_file = os.path.join(params[cf.OUT_DIR], 'azimuth_angle_gamma.tif')
+    shared.write_output_geotiff(md, gt, wkt, azimuth_angle_gamma, azimuth_angle_gamma_file, np.nan)
```

Comment:
not needed for prod code. suggest delete

Reply:
Deleted.